### ISSUE TITLE
Implement IdentityMap::refresh

### ISSUE THREAD

**Issue by hrach** [nextras/orm#23]
In long-run processes it's quite commont we need to refetch actual state of entity in database. This api should be a little bit hidden in IdentityMap. refresh would refetch data and updated entity state. If entity isModified() === TRUE, exception should be thrown.

**Comment by hrach** [comment]
higher priority, I need it now. Yesterday was late :innocent: :angel:

**Comment by hrach** [comment]
Todo:
- [ ] somehow measure gc - ideas?
- [ ] disallow attaching old entities to repository.

**Comment by hrach** [comment]
Advanced implementation postponed after v1.0.0.

**Comment by JanTvrdik** [comment]
:+1: I need this now as well. For thread-safe critical code I'd like to do sth like

~~~php
lock();
$repo->refresh($entity); // make sure I have latest data
doCiritalStuff();
unlock();
~~~

**Comment by JanTvrdik** [comment]
And I need this again.

**Comment by hrach** [comment]
Feel free to open PR. Master is ready to merge it :)

**Comment by hrach** [comment]
Refresh (probably) shouldn't touch relationship properties, it should update only non-relationship keys. Is it ok for your usecase, @JanTvrdik? Updating relationships would have to trigger recursive update, that could end badly.

**Comment by hrach** [comment]
Well, Doctrine do only recursive refresh: https://github.com/doctrine/doctrine2/blob/f88896cc9df756ce2d0e3f3f63bcaf24ef536dd5/lib/Doctrine/ORM/UnitOfWork.php#L2030-L2049

**Comment by hrach** [comment]
non-recurive impl: https://github.com/hrach/nextras-orm/commit/8cd04279d6a8d9f46ccac57e89838c678ccaed70

**Comment by JanTvrdik** [comment]
Yes, non-recursive refresh is currently (hopefully) enough for me.

**Comment by f3l1x** [comment]
@hrach I need it too. 

I have before save trigger in DB and after persist and need to reload data again with new data.

**Comment by Mikulas** [comment]
That would be our use case as well.

Generic refresh aside, this exact refresh could be solved with `INSERT RETURNING` (which returns when *after* triggers are complete). This could be an opt-in feature of some advanced mapper; something like this:

```php
```

**Comment by f3l1x** [comment]
@Mikulas Maybe.

Or just some method to clear entity in identity map.

**Comment by hrach** [comment]
#68 was implemented. Postpoing this to 2.1, since I had to figure out the proper bahavior what to do, when something changes in relationships.

**Comment by JanTvrdik** [comment]
And I need this again.

**Comment by hrach** [comment]
Changing priority since refreshAll will be implemented in 3.0. #219

**Comment by hrach** [comment]
Well, closing for now. #219 & #227.
